### PR TITLE
Add a notepad

### LINKED ISSUE #297: Add private notepad
private notepad/text field if possible, that is not seen by other players in game. Users could record own notes/short term strategy might help remember what we intend to do when game comes back to our turn

### PR DESCRIPTION
**Pull request by jenf**
fixes #297

While this could have been done as local storage, I often switch devices and so losing my notes would be annoying.

(This is partially to implement user storage for #209)

### CODE REVIEW

File: assets/app/view/game/notepad.rb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+        h(:div, [
+          notepad,
+          render_button("Save Notepad#{saved ? '' : ' (Unsaved)'}") { submit },
+          ])
+      end
+
+      def local_save
+        # non-persistently save the notepad in case the user switches tabs
+        store(:user_notes, params['notepad'])
```
Comment (tobymao) [inline assets/app/view/game/notepad.rb:46]: hm this will cause a full rerender every time you type a key stroke?

File: routes/game.rb [hunk 2]
```diff
@@ -36,6 +40,14 @@ class Api
             game.to_h
           end
 
+          # POST '/api/game/<game_id>/user_settings'
+          r.is 'user_settings' do
+            DB.with_advisory_lock(:action_lock, game.id) do
```
Comment (tobymao) [inline routes/game.rb:45]: i don't think this lock is necessary

File: routes/game.rb [hunk 3]
```diff
@@ -11,7 +11,11 @@ class Api
 
         # '/api/game/<game_id>/'
         r.is do
-          game.to_h(include_actions: true)
+          game_data = game.to_h(include_actions: true)
```
Comment (tobymao) [inline routes/game.rb:14]: instead could we make this an option on the to_h, so to_h takes in a param who it's for

File: assets/app/view/game/notepad.rb (hunk 1 above)
Comment (jenf) [reply assets/app/view/game/notepad.rb:46]: On Firefox it's only when you exit the textbox, I'll test with chrome

**Comment by tobymao** [comment]
oh ok cool then

File: routes/game.rb (hunk 2 above)
Comment (jenf) [reply routes/game.rb:45]: If two users try to save the notepad at the same time, won't one user lose their content?

**Comment by tobymao** [comment]
@jenf still waiting on the changes to to_h

**Comment by jenf** [comment]
yeah, i'll pick this up in a few days. Trying to progress 1882 more

File: assets/app/view/game/notepad.rb (hunk 1 above)
Comment (jenf) [reply assets/app/view/game/notepad.rb:46]: Same on Edge and Chrome.

File: routes/game.rb (hunk 2 above)
Comment (jenf) [reply routes/game.rb:43]: @tobymao do you want me to remove the lock? I think it might have race conditions if we remove it.

**Comment by tobymao** [comment]
can keep it

**Pull request merged**